internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -893,6 +925,106 @@ func (r *BMACReconciler) ensureSpokeMachine(ctx context.Context, log logrus.Fiel
 	return machineSpoke, nil
 }
 
+// Get MCS cert of the spoke cluster, encode it and add it to BMH
+func (r *BMACReconciler) ensureMCSCert(ctx context.Context, log logrus.FieldLogger, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent) reconcileResult {
+	// Only worker role is supported for day2 operation
+	if agent.Spec.Role != models.HostRoleWorker || agent.Spec.ClusterDeploymentName == nil {
+		log.Debugf("Skipping spoke BareMetalHost reconcile for  agent %s/%s, role %s and clusterDeployment %s.", agent.Namespace, agent.Name, agent.Spec.Role, agent.Spec.ClusterDeploymentName)
+		return reconcileComplete{dirty: true}
+	}
+
+	cdKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, cdKey, clusterDeployment); err != nil {
+		log.WithError(err).Errorf("failed to get clusterDeployment resource %s/%s", cdKey.Namespace, cdKey.Name)
+		return reconcileError{err}
+	}
+
+	// If the cluster is not installed yet, we can't get kubeconfig for the cluster yet.
+	if !clusterDeployment.Spec.Installed {
+		log.Infof("ClusterDeployment %s/%s for agent %s/%s is not installed yet", clusterDeployment.Namespace, clusterDeployment.Name, agent.Namespace, agent.Name)
+		// If cluster is not Installed, wait until a reconcile is trigged by a ClusterDeployment watch event instead
+		return reconcileComplete{dirty: true}
+	}
+
+	secret := &corev1.Secret{}
+	name := fmt.Sprintf(adminKubeConfigStringTemplate, clusterDeployment.Name)
+
+	err := r.Get(ctx, types.NamespacedName{Namespace: agent.Spec.ClusterDeploymentName.Namespace, Name: name}, secret)
+	if err != nil && k8serrors.IsNotFound(err) {
+		log.WithError(err).Errorf("failed to get secret resource %s/%s", agent.Spec.ClusterDeploymentName.Namespace, name)
+		// TODO: If secret is not found, wait until a reconcile is trigged by a watch event instead
+		return reconcileComplete{dirty: true}
+	} else if err != nil {
+		return reconcileError{err}
+	}
+
+	spokeClient, err := r.getSpokeClient(secret)
+	if err != nil {
+		log.WithError(err).Errorf("failed to create spoke kubeclient")
+		return reconcileError{err}
+	}
+
+	ignitionWithMCSCert, err := r.createIgnitionWithMCSCert(ctx, spokeClient)
```

Comment:
all logic from start of the function till here is the same 100% as in reconcileSpokeBMH. All this code should be move to dedicated function.

Reply:
Please see my comment here https://github.com/openshift/assisted-service/pull/1661/files#r656616130

Reply:
https://github.com/openshift/assisted-service/pull/1661/files#diff-86415614c6ed40c56c5ca2b5f751a65746ef5c83b61e6f021a4f5d817e4425baR1217-R1260